File: .claude/worktrees/nice-bartik-75f638/backend/routers/incidencies.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncConnection

from database import get_db
from auth import get_current_user, require_rrhh_or_admin
from models import IncidenciaIn, IncidenciaStatusIn
from email_service import send_email

router = APIRouter(prefix="/api/incidencies", tags=["incidencies"])


async def _notify(db: AsyncConnection, user_id: int, title: str, body: str) -> None:
    await db.execute(
        text("INSERT INTO notifications (user_id, title, body, tab) VALUES (:uid, :title, :body, 'Veu/Incidències')"),
        {"uid": user_id, "title": title, "body": body},
    )


@router.get("")
async def list_incidencies(
    current_user: dict = Depends(get_current_user),
    db: AsyncConnection = Depends(get_db),
):
    rows = (await db.execute(
        text("SELECT * FROM incidencies ORDER BY created_at DESC")
    )).mappings().all()
    return [dict(r) for r in rows]


@router.post("", status_code=201)
async def create_incidencia(
    body: IncidenciaIn,
    current_user: dict = Depends(get_current_user),
    db: AsyncConnection = Depends(get_db),
):
    result = await db.execute(
        text("""
            INSERT INTO incidencies (title, description, area, priority, author, user_id)
            VALUES (:title, :description, :area, :priority, :author, :user_id)
        """),
        {"title": body.title, "description": body.description,
         "area": body.area, "priority": body.priority, "author": current_user["name"],
         "user_id": current_user["id"]},
    )

    # Notify and email all "Manteniment" users
    maint_users = (await db.execute(
        text("SELECT id, email, email_notifs FROM users WHERE role = 'Manteniment'")
    )).mappings().all()

    for u in maint_users:
        await _notify(
            db, u["id"],
            f"Nova incidència: {body.title}",
            f"{current_user['name']} ha registrat una incidència a {body.area} (prioritat: {body.priority}).",
        )
        if u["email_notifs"]:
            await send_email(
                to=u["email"],
                subject=f"Nova incidència: {body.title}",
                html=f"""
                <p>Hola,</p>
                <p>S'ha registrat una nova incidència al portal TAVIL:</p>
                <ul>
                  <li><b>Títol:</b> {body.title}</li>
                  <li><b>Àrea:</b> {body.area}</li>
                  <li><b>Prioritat:</b> {body.priority}</li>
                  <li><b>Descripció:</b> {body.description}</li>
                  <li><b>Reportada per:</b> {current_user['name']}</li>
                </ul>
                <p>Accedeix al portal per gestionar-la.</p>
                """,
            )

    await db.commit()
    row = (await db.execute(
        text("SELECT * FROM incidencies WHERE id = :id"), {"id": result.lastrowid}
    )).mappings().first()
    return dict(row)


@router.patch("/{incidencia_id}/status")
async def update_status(
    incidencia_id: int,
    body: IncidenciaStatusIn,
    _admin: dict = Depends(require_rrhh_or_admin),
    db: AsyncConnection = Depends(get_db),
):
    await db.execute(
        text("""
            UPDATE incidencies
            SET status = :status, assigned_to = :assigned_to, resolution = :resolution
            WHERE id = :id
        """),
        {"status": body.status, "assigned_to": body.assigned_to,
         "resolution": body.resolution, "id": incidencia_id},
    )
    await db.commit()
    return {"ok": True}


@router.delete("/{incidencia_id}", status_code=204)
async def delete_incidencia(
    incidencia_id: int,
    current_user: dict = Depends(get_current_user),
    db: AsyncConnection = Depends(get_db),
):
    row = (await db.execute(
        text("SELECT user_id FROM incidencies WHERE id = :id"), {"id": incidencia_id}
    )).mappings().first()
    if not row:
        raise HTTPException(status_code=404, detail="No trobat")
    is_admin = current_user["role"] in ("Administrador/a", "Recursos humans")
    if not is_admin and row["user_id"] != current_user["id"]:
        raise HTTPException(status_code=403, detail="No autoritzat")
    await db.execute(text("DELETE FROM incidencies WHERE id = :id"), {"id": incidencia_id})
    await db.commit()

File: api/routes/incidencies.php
<?php
// Routes: /api/incidencies/*
require_once __DIR__ . '/../db.php';
require_once __DIR__ . '/../auth_middleware.php';
require_once __DIR__ . '/../email.php';
require_once __DIR__ . '/../helpers.php';

$db   = get_db();
$body = request_body();
$id   = isset($segments[1]) && is_numeric($segments[1]) ? (int)$segments[1] : null;
$sub  = $segments[2] ?? '';

// GET /api/incidencies
if ($method === 'GET' && $id === null) {
    $u = auth_user();
    $is_privileged = user_has_any_role($u, ['Administrador', 'Administrador/a', 'Recursos humans', 'SolicitudsVacances', 'SolicitudsDissabtes']);
    if ($is_privileged) {
        $stmt = $db->query('SELECT * FROM incidencies ORDER BY created_at DESC');
    } else {
        $stmt = $db->prepare('SELECT * FROM incidencies WHERE author=? ORDER BY created_at DESC');
        $stmt->execute([$u['name']]);
    }
    $rows = $stmt->fetchAll();
    foreach ($rows as &$r) $r['id'] = (int)$r['id'];
    respond($rows);
}

// POST /api/incidencies
elseif ($method === 'POST' && $id === null) {
    $u = auth_user();
    $db->prepare('INSERT INTO incidencies (title,description,area,priority,author) VALUES (?,?,?,?,?)')
       ->execute([str_val($body,'title'), str_val($body,'description'), str_val($body,'area','General'), str_val($body,'priority','Baixa'), $u['name']]);
    $new_id = (int)$db->lastInsertId();

    // Notify Manteniment users
    $maint = $db->query("SELECT id,email,email_notifs FROM users WHERE role='Manteniment'")->fetchAll();
    foreach ($maint as $m) {
        push_notification($db, (int)$m['id'],
            'Nova incidència: ' . str_val($body,'title'),
            $u['name'] . ' ha registrat una incidència a ' . str_val($body,'area') . ' (prioritat: ' . str_val($body,'priority') . ').',
            'Veu/Incidències'
        );
        if ($m['email_notifs']) {
            send_email($m['email'], 'Nova incidència: ' . str_val($body,'title'),
                "<p>Hola,</p><p>S'ha registrat una nova incidència al portal TAVIL:</p><ul>"
                . '<li><b>Títol:</b> ' . htmlspecialchars(str_val($body,'title')) . '</li>'
                . '<li><b>Àrea:</b> ' . htmlspecialchars(str_val($body,'area')) . '</li>'
                . '<li><b>Prioritat:</b> ' . htmlspecialchars(str_val($body,'priority')) . '</li>'
                . '<li><b>Descripció:</b> ' . htmlspecialchars(str_val($body,'description')) . '</li>'
                . '<li><b>Reportada per:</b> ' . htmlspecialchars($u['name']) . '</li></ul>'
                . '<p>Accedeix al portal per gestionar-la.</p>'
            );
        }
    }

    $row = $db->query("SELECT * FROM incidencies WHERE id=$new_id")->fetch();
    $row['id'] = (int)$row['id'];
    respond($row, 201);
}

// PATCH /api/incidencies/{id}/status
elseif ($method === 'PATCH' && $id !== null && $sub === 'status') {
    require_rrhh_or_admin();
    $db->prepare('UPDATE incidencies SET status=?, assigned_to=?, resolution=? WHERE id=?')
       ->execute([str_val($body,'status'), str_val($body,'assigned_to'), str_val($body,'resolution'), $id]);
    respond(['ok' => true]);
}

else {
    respond(['detail' => 'Not found'], 404);
}
